### PR TITLE
SDI-76 New endpoint to retrieve list of domains

### PR DESCRIPTION
**Pull request by andymarke**
Endpoint is likely to only be used by developers who want to guess at the domain they need for other APIs

- Use JPA Repository for domain rather then SQL version since this is standard going forward
- Use JsonPath for testing rather than json file assert that is confusing

### CODE REVIEW

**Review by lbennett-moj — APPROVED** [review]
LGTM

File: src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/ReferenceDomain.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+import javax.persistence.Entity;
+import javax.persistence.Id;
+import javax.persistence.Table;
+
+@AllArgsConstructor
+@NoArgsConstructor
+@Getter
+@Entity
+@Table(name = "reference_domains")
+public class ReferenceDomain {
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/ReferenceDomain.java:17]: new classes should be kotlin :grin:

File: src/test/java/uk/gov/justice/hmpps/prison/api/resource/impl/ReferenceDataResourceTest.java [hunk 2]
```diff
@@ -181,4 +185,74 @@ public void testReadDomainReverseLookupWithWildcard() {
...
+
+        @Test
+        @DisplayName("will return a list of domains")
+        void willReturnAListOfDomains() {
+            final var token = authTokenHelper.getToken(AuthTokenHelper.AuthToken.NORMAL_USER);
+
+            final var response = getBodyAsJsonContent(getDomains(token));
+
+            assertThat(response).hasJsonPathArrayValue("$");
+            assertThat(response).extractingJsonPathNumberValue("$.length()").isEqualTo(430);
```
Comment (petergphillips) [inline src/test/java/uk/gov/justice/hmpps/prison/api/resource/impl/ReferenceDataResourceTest.java:213]: dubious about this one - looks like the test will be fragile as any new domain will cause it to fail.  Could you do a `> 300` and `< 600` instead?

File: src/test/java/uk/gov/justice/hmpps/prison/api/resource/impl/ReferenceDataResourceTest.java [hunk 3]
```diff
@@ -181,4 +185,74 @@ public void testReadDomainReverseLookupWithWildcard() {
...
+
+            assertThat(response)
+                .hasJsonPathValue(domainAt, "ETHNICITY");
+            assertThat(response)
+                .hasJsonPathValue(domainAt, "SKL_SUB_TYPE");
+
+            assertThat(response)
+                .extractingJsonPathValue(domainPropertyAt, "ETHNICITY", "description").asList().element(0)
+                .isEqualTo("Ethnicity");
+            assertThat(response)
```
[the remark this answers is not in the document]
Comment (petergphillips) [reply src/test/java/uk/gov/justice/hmpps/prison/api/resource/impl/ReferenceDataResourceTest.java:234]: you can use `assertThatJson` instead in this project.  But I agree, we should just add webflux and be done with it

**Pull request merged**